### PR TITLE
feat(resharding): Full forks support in flat storage

### PR DESCRIPTION
**Pull request by Trisfald**
This PR fixes the remaining issue in `FlatStorageResharder` that prevented the chain from progressing at resharding boundary in some corner cases.

Two possible solutions have been considered: 
- Postponing the split task until the last block of the epoch becomes final
- Postponing the split task until one candidate resharding block becomes final 

I've decided to proceed with the second option, as it was building on existing code, and the changes were limited in scope to the inner impl of `FlatStorageResharder`. In this way the resharder remains entirely decoupled from epoch manager.

Also in this PR:
- Minor refactoring: task scheduling became a bit more involved in order to handle multiple resharding requests that can get intertwined.
- New unit test to validate the implementation

Now the existing tests for forks in `resharding_v3` have been finally `unignored` 💪

### CODE REVIEW

File: chain/chain/src/flat_storage_resharder.rs [hunk 1]
```diff
@@ -181,6 +181,22 @@ impl FlatStorageResharder {
         info!(target: "resharding", ?split_params, "initiating flat storage shard split");
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:181]: Can you add a comment to the method about the fork handling? 
github doesn't let me comment there but I mean split_shard a few lines above here.

File: chain/chain/src/flat_storage_resharder.rs [hunk 2]
```diff
@@ -181,6 +181,22 @@ impl FlatStorageResharder {
...
+        // Get all resharding blocks, if any, from the existing scheduled split shard,
+        // and merge them with the new resharding block.
+        let mut resharding_blocks = vec![resharding_block];
+        if let Some(FlatStorageReshardingEventStatus::SplitShard(parent, params, _)) =
+            self.resharding_event()
+        {
+            if parent_shard == parent
+                && params.left_child_shard == left_child_shard
+                && params.right_child_shard == right_child_shard
+                && params.shard_layout == *shard_layout
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:194]: Why do you need those checks?

File: chain/chain/src/flat_storage_resharder.rs [hunk 3]
```diff
@@ -314,63 +310,111 @@ impl FlatStorageResharder {
...
+    /// Checks if the resharder is ready to split a shard. In such case it returns the parameters needed for the split.
+    fn split_shard_check_readiness(&self, chain_store: &ChainStore) -> SplitShardSchedulingStatus {
+        info!(target: "resharding", "checking readiness of flat storage shard split task");
+        // The current resharding event must exist and be compatible with the operation of
+        // splitting a shard. All these checks must happen while the resharding event is under
+        // lock, to prevent multiple parallel resharding attempts (due to forks) to interfere
+        // with each other .
+        let mut event_lock_guard = self.resharding_event.lock().unwrap();
+        let (resharding_blocks, parent_shard, split_params, execution_status) = if let Some(event) =
+            event_lock_guard.as_mut()
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:363]: nit: Would the following work? 
```
let Some(event) = event_lock_guard.as_mut() else {
  info!(target: "resharding", "flat storage shard split task cancelled: no resharding event");
  return SplitShardSchedulingStatus::Cancelled;
}
...
```

File: chain/chain/src/flat_storage_resharder.rs [hunk 4]
```diff
@@ -314,63 +310,111 @@ impl FlatStorageResharder {
...
+        {
+            if event.has_started() {
+                info!(target: "resharding", "flat storage shard split task cancelled: resharding already in progress");
+                return SplitShardSchedulingStatus::Cancelled;
+            }
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:368]: Is this expected to happen under normal circumstances, under fork or under error?

File: chain/chain/src/flat_storage_resharder.rs [hunk 5]
```diff
@@ -847,20 +887,45 @@ impl FlatStorageResharder {
...
+        resharding_blocks: &[BlockInfo],
         chain_store: &ChainStore,
-    ) -> FlatStorageReshardingTaskSchedulingStatus {
+    ) -> TaskSchedulingStatus {
+        use TaskSchedulingStatus::*;
+        let mut at_least_one_postponed = false;
+        for block_info in resharding_blocks {
+            match self.compute_scheduled_task_status_at_block(&block_info, chain_store) {
+                CanStart(block_info) => return CanStart(block_info),
+                Failed => {} // Nothing to do on failure.
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:903]: Can you mention that this may happen during forks and isn't actually "the end of the world"?

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 6]
```diff
@@ -726,12 +724,10 @@ fn test_resharding_v3_resharding_block_in_fork() {
 }
 
 #[test]
-// TODO(resharding): fix nearcore and un-ignore this test
 // TODO(resharding): duplicate this test so that in one case resharding is performed on block
 //                   B(height=13) and in another case resharding is performed on block B'(height=13).
 //                   In the current scenario the real resharding happens on block B'. Low priority TODO
 //                   since it's a very rare corner case.
```
Comment (wacban) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:730]: Is this todo still to do?

**Review by wacban — APPROVED** [review]
LGTM

**Comment by wacban** [comment]
> Two possible solutions have been considered:
    Postponing the split task until the last block of the epoch becomes final
    Postponing the split task until one candidate resharding block becomes final

Is the first option worth considering as a better engineering or good first issue project for later?

File: chain/chain/src/flat_storage_resharder.rs (hunk 2 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:194]: These are not strictly necessary. The checks are a safeguard to discard the old candidate resharding block if the shard layout of the new resharding is different. It can't really happen today since shard layout is fixed.

File: chain/chain/src/flat_storage_resharder.rs (hunk 1 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:181]: Sure

File: chain/chain/src/flat_storage_resharder.rs (hunk 3 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:363]: Yep, thanks!

File: integration-tests/src/test_loop/tests/resharding_v3.rs (hunk 6 above)
Comment (Trisfald) [reply integration-tests/src/test_loop/tests/resharding_v3.rs:730]: Yes this must be done still. `split_shard_in_forks_when_older_branch_is_finalized` is doing that but only for flat storage unit testing; it would be nice to have the test loop version with all the components.

File: chain/chain/src/flat_storage_resharder.rs (hunk 4 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:368]: Only if there are forks involved

**Comment by Trisfald** [comment]
> > Two possible solutions have been considered:
> > Postponing the split task until the last block of the epoch becomes final
> > Postponing the split task until one candidate resharding block becomes final
> 
> Is the first option worth considering as a better engineering or good first issue project for later?

It's an alternative with different tradeoffs. It makes easier handling the state, since instead of block hashes we just need one epoch_id. On the other hand, it makes resharder less decoupled because of the dependency on epochs. 
Complexity wise - I expect both to be on a similar level, since dealing with blocks is somewhat easier than dealing with epochs. 

TL;DR Not a straight, simple improvement IMO
